### PR TITLE
WIP: Multiple token standard support

### PR DESCRIPTION
**Pull request by mxmauro**

### CODE REVIEW

**Comment by mxmauro** [comment]
Hi @jijordre , here the list of modifications:

1. Implemented `BalanceLib` to handle all kind of balances and tracking history of deposits and withdrawals.
2. Reordered parameters of `TokenController`'s `send` and `receive`. (noticed still missing to reorder `amount` by the time I wrote this)
3. `TokenManager` interface not accepts a string as the "standard". The idea is to initiallty map "erc20" to `ERC20TokenController` and "erc721" to `ERC721TokenController`. So a user can override the controller by specifying human-friendly string rather than an ID.
4. `getTokenController` accepts the standard string if specified BUT if the token is already registered, the registration mapping will take precedence.
5. At last you can see initial changes to `ClientFund`. With the help of `BalanceLib` the code is lot cleaner.

File: contracts/BalanceLib.sol [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+import {SafeMathInt} from "./SafeMathInt.sol";
+
+library BalanceLib {
+    using SafeMathInt for int256;
+
+    struct Deposit {
+        int256 amount;
+        uint256 timestamp;
+        address token;      //0 for ethers
+        uint256 nonfungible_id;
```
Comment (jijordre) [inline contracts/BalanceLib.sol]: As ERC1155 uses ID for both FT and NFT I suggest that properties and params `nonfungible_id` be replaced by `currencyId` (implying also a shift from snake case to camel case), or possibly as simple as `id` as that is used in `BalanceLib`.

In addition as ERC1155 supports multiple tokens per contract I suggest that property `token` be replaced by `contract`.

File: contracts/BalanceLib.sol [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    struct Withdrawal {
+        int256 amount;
+        uint256 timestamp;
+        address token;      //0 for ethers
+        uint256 nonfungible_id;
+    }
+
+    struct Balance {
+        int256 ethers;
+        mapping(address => mapping(uint256 => int256)) tokens;
```
Comment (jijordre) [inline contracts/BalanceLib.sol]: In light of my comment above wouldn't it be possible to remove Balance' `ethers` and rather store ether values in Balance' `tokens[address(0)][0]`, also suggested by #125.

File: contracts/TokenManager.sol [hunk 3]
```diff
@@ -69,49 +77,63 @@ contract TokenManager is Ownable, Modifiable, SelfDestructible {
...
+        if (registeredTokens[token].interface_id != bytes32(0)) {
+            require(!registeredTokens[token].blacklisted);
+
+            address interface_address = registeredTokenControllers[registeredTokens[token].interface_id];
+            require(interface_address != address(0));
+
+            return TokenController(interface_address);
+        }
+
+        if (bytes(standard).length > 0) {
```
Comment (jijordre) [inline contracts/TokenManager.sol]: Here I don't think that any existing mapping should override `standard`. In the event that the mapping of token to interface ID is wrong the given `standard` should take precedence.

Comment (mxmauro) [reply contracts/TokenManager.sol]: Fine! I'll switch them

File: contracts/BalanceLib.sol (hunk 2 above)
Comment (mxmauro) [reply contracts/BalanceLib.sol]: I agree

File: contracts/BalanceLib.sol (hunk 1 above)
Comment (mxmauro) [reply contracts/BalanceLib.sol]: 👍

File: contracts/ClientFund.sol [hunk 4]
```diff
@@ -79,8 +82,17 @@ contract ClientFund is Ownable, Modifiable, Beneficiary, Benefactor, Authorizabl
         serviceActivationTimeout = 1 weeks;
     }
 
-    function setCurrencyManager(address manager) public onlyOwner {
-        currencyManager = CurrencyManager(manager);
+    function setCurrencyManagerAddress(address newAddress) public onlyOwner notNullAddress(newAddress) {
```
Comment (jijordre) [inline contracts/ClientFund.sol]: I suggest this function signature be changed to `changeCurrencyManager(CurrencyManager newCurrencyManager) public` in order to fit the pattern used in other contracts. Alternatively we could also replace "change" with "set" in function that change auxiliary contract address in the full set contracts.

File: contracts/TokenHolderRevenueFund.sol [hunk 5]
```diff
@@ -96,18 +90,31 @@ contract TokenHolderRevenueFund is Ownable, AccrualBeneficiary, Servable, SelfDe
     // Functions
     // -----------------------------------------------------------------------------------------------------------------
     function setRevenueTokenAddress(address newAddress) public onlyOwner notNullAddress(newAddress) {
```
Comment (jijordre) [inline contracts/TokenHolderRevenueFund.sol]: Although not changed in this PR I suggest this function signature be changed to `changeRevenueToken(RevenueToken newRevenueToken) public` in order to fit the pattern used in other contracts.

File: contracts/TokenHolderRevenueFund.sol [hunk 6]
```diff
@@ -96,18 +90,31 @@ contract TokenHolderRevenueFund is Ownable, AccrualBeneficiary, Servable, SelfDe
...
-            revenueToken = newAddress;
+            revenueToken = RevenueToken(newAddress);
 
-            // Emit event
+            //emit event
             emit RevenueTokenChangedEvent(oldAddress, newAddress);
         }
     }
 
+    function setCurrencyManagerAddress(address newAddress) public onlyOwner notNullAddress(newAddress) {
```
Comment (jijordre) [inline contracts/TokenHolderRevenueFund.sol]: Although not changed in this PR I suggest this function signature be changed to `changeCurrencyManager(CurrencyManager newCurrencyManager) public` in order to fit the pattern used in other contracts.

File: contracts/ClientFundable.sol [hunk 7]
```diff
@@ -28,21 +26,22 @@ contract ClientFundable is Ownable, Modifiable {
...
-    function changeClientFund(ClientFund newClientFund)
-    public
-    onlyOwner
-    notNullAddress(newClientFund)
-    {
-        ClientFund oldClientFund = clientFund;
-        clientFund = newClientFund;
-        emit ChangeClientFundEvent(oldClientFund, clientFund);
+    /// @param newAddress The (address of) ClientFund contract instance
+    function changeClientFund(address newAddress) public onlyOwner notNullAddress(newAddress) {
```
Comment (jijordre) [inline contracts/ClientFundable.sol]: What is the reason for this change?

File: contracts/ClientFundable.sol [hunk 8]
```diff
@@ -28,21 +26,22 @@ contract ClientFundable is Ownable, Modifiable {
...
+        if (newAddress != address(clientFund)) {
+            //set new community vote
+            address oldAddress = address(clientFund);
+            communityVote = ClientFund(newAddress);
```
Comment (jijordre) [inline contracts/ClientFundable.sol]: Has to be `clientFund` field instead of `communityVote`.

File: contracts/CommunityVotable.sol [hunk 9]
```diff
@@ -39,16 +37,18 @@ contract CommunityVotable is Ownable, Modifiable {
...
 
     /// @notice Change the community vote contract
-    /// @param newCommunityVote The (address of) CommunityVote contract instance
-    function changeCommunityVote(CommunityVote newCommunityVote)
-    public
-    onlyOwner
-    notNullAddress(newCommunityVote)
-    {
+    /// @param newAddress The (address of) CommunityVote contract instance
+    function changeCommunityVote(address newAddress) public onlyOwner notNullAddress(newAddress) {
```
Comment (jijordre) [inline contracts/CommunityVotable.sol]: Idem regarding reason for change.

File: contracts/Configurable.sol [hunk 10]
```diff
@@ -26,18 +24,22 @@ contract Configurable is Ownable, Modifiable {
...
-    function changeConfiguration(Configuration newConfiguration)
-    public
-    onlyOwner
-    notNullAddress(newConfiguration)
-    {
-        Configuration oldConfiguration = configuration;
-        configuration = newConfiguration;
-        emit ChangeConfigurationEvent(oldConfiguration, configuration);
+    /// @param newAddress The (address of) Configuration contract instance
+    function changeConfiguration(address newAddress) public onlyOwner notNullAddress(newAddress) {
```
Comment (jijordre) [inline contracts/Configurable.sol]: Idem regarding reason for change.

File: contracts/FraudChallengable.sol [hunk 11]
```diff
@@ -33,17 +31,21 @@ contract FraudChallengable is Ownable, Modifiable, SelfDestructible {
...
-    function changeFraudChallenge(FraudChallenge newFraudChallenge)
-    public
-    onlyOwner
-    notNullAddress(newFraudChallenge)
-    {
-        FraudChallenge oldFraudChallenge = fraudChallenge;
-        fraudChallenge = newFraudChallenge;
-        emit ChangeFraudChallengeEvent(oldFraudChallenge, fraudChallenge);
+    /// @param newAddress The (address of) FraudChallenge contract instance
+    function changeFraudChallenge(address newAddress) public onlyOwner notNullAddress(newFraudChallenge) {
```
Comment (jijordre) [inline contracts/FraudChallengable.sol]: Idem regarding reason for change.

File: contracts/Hashable.sol [hunk 12]
```diff
@@ -28,17 +25,18 @@ contract Hashable is Ownable, Modifiable {
...
     /// @param newHasher The (address of) Hasher contract instance
-    function changeHasher(Hasher newHasher)
-    public
-    onlyOwner
-    notNullAddress(newHasher)
-    {
-        Hasher oldHasher = hasher;
-        hasher = newHasher;
-        emit ChangeHasherEvent(oldHasher, hasher);
+    function changeHasher(address newAddress) public onlyOwner notNullAddress(newAddress) {
```
Comment (jijordre) [inline contracts/Hashable.sol]: Idem regarding reason for change.

File: contracts/SecurityBondable.sol [hunk 13]
```diff
@@ -26,20 +24,21 @@ contract SecurityBondable is Ownable, Modifiable {
...
-    function changeSecurityBond(SecurityBond newSecurityBond)
-    public
-    onlyOwner
-    notNullAddress(newSecurityBond)
-    {
-        SecurityBond oldSecurityBond = securityBond;
-        securityBond = newSecurityBond;
-        emit ChangeSecurityBondEvent(oldSecurityBond, securityBond);
+    /// @param newAddress The (address of) SecurityBond contract instance
+    function changeSecurityBond(address newAddress) public onlyOwner notNullAddress(newAddress) {
```
Comment (jijordre) [inline contracts/SecurityBondable.sol]: Idem regarding reason for change.

File: contracts/Validatable.sol [hunk 14]
```diff
@@ -28,21 +25,22 @@ contract Validatable is Ownable, Modifiable {
...
     /// @param newValidator The (address of) Validator contract instance
-    function changeValidator(Validator newValidator)
-    public
-    onlyOwner
-    notNullAddress(newValidator)
-    {
-        Validator oldValidator = validator;
-        validator = newValidator;
-        emit ChangeValidatorEvent(oldValidator, validator);
+    function changeValidator(address newAddress) public onlyOwner notNullAddress(newAddress) {
```
Comment (jijordre) [inline contracts/Validatable.sol]: Idem regarding reason for change.
